File: src/test/java/pl/bratek20/keyvaluestorage/fixtures/Assert.java
package pl.bratek20.keyvaluestorage.fixtures;

import pl.bratek20.keyvaluestorage.api.KeyValuePair;
import pl.bratek20.keyvaluestorage.api.Value;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class Assert {
    public static void assertValue(Value given, String expected) {
        assertEquals(given.value(), expected);
    }

    public record ExpectedKeyValuePair(String key, String value) { }
    private static void assertKeyValuePair(KeyValuePair given, ExpectedKeyValuePair expected) {
        assertEquals(given.key().value(), expected.key());
        assertValue(given.value(), expected.value());
    }

    public static void assertKeyValuePairList(List<KeyValuePair> given, List<ExpectedKeyValuePair> expected) {
        assertEquals(given.size(), expected.size());
        for (var i = 0; i < given.size(); i++) {
            assertKeyValuePair(given.get(i), expected.get(i));
        }
    }
}

File: src/test/java/pl/bratek20/keyvaluestorage/fixtures/Assert.kt
package pl.bratek20.keyvaluestorage.fixtures

import org.assertj.core.api.Assertions.assertThat
import pl.bratek20.keyvaluestorage.api.Key
import pl.bratek20.keyvaluestorage.api.KeyValuePair
import pl.bratek20.keyvaluestorage.api.Value

object Assert {
    fun value(given: Value, expected: String) {
        assertThat(given.value).isEqualTo(expected)
    }

    private fun keyValuePair(given: KeyValuePair, expectedConfigure: ExpectedKeyValuePair.() -> Unit) {
        val expected = ExpectedKeyValuePair().apply(expectedConfigure)
        if (expected.key != null) {
            assertThat(given.key.value).isEqualTo(expected.key)
        }
        if (expected.value != null) {
            assertThat(given.value.value).isEqualTo(expected.value)
        }
    }

    fun keyValuePairList(given: List<KeyValuePair>, expected: List<ExpectedKeyValuePair.() -> Unit>) {
        assertThat(given).hasSize(expected.size)
        given.forEachIndexed { index, keyValuePair ->
            keyValuePair(keyValuePair, expected[index])
        }
    }

    fun keyList(given: List<Key>, expected: List<String>) {
        assertThat(given).hasSize(expected.size)
        given.forEachIndexed { index, key ->
            assertThat(key.value).isEqualTo(expected[index])
        }
    }

    class ExpectedKeyValuePair(var key: String? = null, var value: String? = null)
}
